### ISSUE TITLE
Slurm Profile fails to detect or remove incomplete output from cancelled jobs

### ISSUE THREAD

**Issue by johnstonmj** [Snakemake-Profiles/slurm#49]
After interrupting a job that is running via the slurm profile, snakemake fails to detect incomplete output.

Setup:
```
# Reset
rm -r logs/ output/ input/
# Toy input
mkdir -p logs/long_running_rule
mkdir -p input
touch input/{apple,banana,cucumber}.txt
```

minimal_example.smk
```
rule all:
    input: expand("output/{sample}.delayOutput.txt", sample = ["apple", "banana", "cucumber"])

rule long_running_rule:
    input:
        "input/{sample}.txt"
    output:
        "output/{sample}.delayOutput.txt",
    resources:
        mem = 1000,
        time = 10,
    shell:
        """
        echo "" > {output}

        for i in $(seq 1 60)
        do
            echo $i >> {output}
            sleep 1
        done
        """
```

### Desired behavior: Local job
`snakemake --snakefile "minimal_example.smk" --cores 3 all`
Interrupt local job while running by using:
`chmod u-w output/apple.delayOutput.txt`
Snakemake job manager detects failed job
Critically includes: "Removing output files of failed job long_running_rule since they might be corrupted"
Subsequent snakemake call correctly tries to re-run this job

### Bug: Job submission via slurm profile
```
snakemake --snakefile "minimal_example.smk" --profile slurm --cluster-config config/cluster_config.yaml
```
Interrupt job using:
`scancel`
Snakemake job manager detects incomplete job 
Output file is **not removed nor marked as incomplete**
Subsequent snakemake call incorrectly states 'Nothing to be done.' even though the file is truncated

### Fails but differently: Cluster job submission without profile
```
snakemake --snakefile "minimal_example.smk" --latency-wait 60 --jobs 10 --cluster "sbatch -N 1 -n 1 --mem=20 -t 00:10:00 -o logs/{rule}/{rule}.{wildcards}.out -e logs/{rule}/{rule}.{wildcards}.err < " all
```
Interrupt job using:
`scancel`
Snakemake job manager **fails to detect incomplete job**
Manual interrupt snakemake job manager with Ctrl+C
Subsequent snakemake call indicates: IncompleteFilesException 
Warns to use: --cleanup-metadata or --rerun-incomplete

Is there a better way to detect and handle incomplete jobs, especially those that produce a partial output file?
Thanks!

**Comment by percyfal** [comment]
Hi,
I started looking at this yesterday and I've done some trials and testing of code. I realized that you can reproduce this issue just by letting a job timeout. The test suite has such a case, but the output file wasn't generated (presumably buffered) and the rule didn't fail. By adding output similar to your case (loop and print integers) I can reproduce the issue! Incomplete files remain and causes snakemake to think the job is done. I have documented my code back-tracking efforts below for future reference and in case I've been side-tracked, as ultimately I think this is an upstream issue with snakemake and not the profile itself.

As alluded to, there is not much the slurm profile can do, but the error is rather due to how the [GenericClusterExecutor](https://github.com/snakemake/snakemake/blob/07f73006dcee1e6ce6c271355fff83705f188de4/snakemake/executors/__init__.py#L809) is implemented. When a job is cancelled with `scancel`, snakemake handles the interrupt by calling [ClusterExecutor.handle_job_error](https://github.com/snakemake/snakemake/blob/07f73006dcee1e6ce6c271355fff83705f188de4/snakemake/executors/__init__.py#L775), which in turn calls [Persistence.cleanup](https://github.com/snakemake/snakemake/blob/07f73006dcee1e6ce6c271355fff83705f188de4/snakemake/persistence.py#L201) to perform job cleanup. The latter removes the job metadata, but AFAICT does nothing about the incomplete file, hence the error. I tested by adding `job.cleanup()` to Persistence.cleanup and the issue was resolved. However, there may be a reason this step is omitted but we'd have to discuss this on the snakemake github pages.

With regards to Ctrl-C, only the drmaa executor currently handles these cases properly. See [DRMAAExecutor.cancel()](https://github.com/snakemake/snakemake/blob/07f73006dcee1e6ce6c271355fff83705f188de4/snakemake/executors/__init__.py#L1211) and compare to [GenericClusterExecutor.cancel](https://github.com/snakemake/snakemake/blob/07f73006dcee1e6ce6c271355fff83705f188de4/snakemake/executors/__init__.py#L866); the latter simply calls a shutdown function ([ClusterExecutor.shutdown](https://github.com/snakemake/snakemake/blob/07f73006dcee1e6ce6c271355fff83705f188de4/snakemake/executors/__init__.py#L659)) and does not utilize any information about jobids so it can't know what files to remove. No doubt this could be addressed with a specialized SlurmExecutor but that is outside of the scope of what I can do anytime soon.

To summarize, stick with scancel and let's push an issue upstream to add removal of incomplete files. There is a similar issue up already (https://github.com/snakemake/snakemake/issues/570) which I'll comment on to keep the discussion alive.

Thanks for the report!

Cheers, Per

**Comment by percyfal** [comment]
This has been addressed here: https://github.com/snakemake/snakemake/pull/857
